federation/pkg/federation-controller/service/service_controller_test.go

```diff
@@ -17,69 +17,268 @@ limitations under the License.
+	c1ServiceWatch.Delete(service)
+	assert.NoError(t, WaitForClusterService(t, sc.serviceFederatedInformer.GetTargetStore(), cluster1.Name,
+		key, service, wait.ForeverTestTimeout))
+
+	close(stop)
+}
+
+func NewService(name string, port int32) *apiv1.Service {
+	return &apiv1.Service{
+		ObjectMeta: apiv1.ObjectMeta{
+			Name:      name,
+			Namespace: apiv1.NamespaceDefault,
+			SelfLink:  "/api/v1/namespaces/default/services/" + name,
+			Labels:    map[string]string{"app": name},
+		},
+		Spec: apiv1.ServiceSpec{
+			Ports: []apiv1.ServicePort{{Port: port}},
+			Type:  apiv1.ServiceTypeLoadBalancer,
 		},
-		knownClusterSet: make(sets.String),
 	}
+}
 
-	tests := []struct {
-		cluster           v1beta1.Cluster
-		expectAccept      bool
-		name              string
-		serviceController *ServiceController
-	}{
-		{
-			cluster:           v1beta1.Cluster{},
-			expectAccept:      false,
-			name:              "empty",
-			serviceController: &serviceController,
+func NewEndpoint(name, ip string) *apiv1.Endpoints {
+	return &apiv1.Endpoints{
+		ObjectMeta: apiv1.ObjectMeta{
+			Name:      name,
+			Namespace: apiv1.NamespaceDefault,
+			SelfLink:  "/api/v1/namespaces/default/endpoints/" + name,
+			Labels:    map[string]string{"app": name},
 		},
-		{
-			cluster: v1beta1.Cluster{
-				Status: v1beta1.ClusterStatus{
-					Conditions: []v1beta1.ClusterCondition{
-						{Type: v1beta1.ClusterReady, Status: v1.ConditionTrue},
-					},
-				},
-			},
-			expectAccept:      true,
-			name:              "basic",
-			serviceController: &serviceController,
+		Subsets: []apiv1.EndpointSubset{{
+			Addresses: []apiv1.EndpointAddress{{
+				IP: ip,
+			}}},
+		},
+	}
+}
+
+// NewCluster builds a new cluster object.
+func NewClusterWithRegionZone(name string, readyStatus apiv1.ConditionStatus, region, zone string) *fedv1.Cluster {
+	return &fedv1.Cluster{
+		ObjectMeta: apiv1.ObjectMeta{
+			Name:        name,
+			Annotations: map[string]string{},
 		},
-		{
-			cluster: v1beta1.Cluster{
-				Status: v1beta1.ClusterStatus{
-					Conditions: []v1beta1.ClusterCondition{
-						{Type: v1beta1.ClusterReady, Status: v1.ConditionFalse},
-					},
-				},
+		Status: fedv1.ClusterStatus{
+			Conditions: []fedv1.ClusterCondition{
+				{Type: fedv1.ClusterReady, Status: readyStatus},
 			},
-			expectAccept:      false,
-			name:              "notready",
-			serviceController: &serviceController,
+			Region: region,
+			Zones:  []string{zone},
 		},
 	}
-	pred := getClusterConditionPredicate()
-	for _, test := range tests {
-		accept := pred(test.cluster)
-		if accept != test.expectAccept {
-			t.Errorf("Test failed for %s, expected %v, saw %v", test.name, test.expectAccept, accept)
+}
+
+// Wait for federation service status to be updated to match the desiredStatus.
+func WaitForClusterService(t *testing.T, store util.FederatedReadOnlyStore, clusterName, key string, desiredService *apiv1.Service, timeout time.Duration) error {
+	retryInterval := 100 * time.Millisecond
```

Comment:
This value is used in multiple places in this test.  Rather lift it into a common named constant.